File: backend/password_policy.py
"""Shared password complexity rules for credential updates."""

from __future__ import annotations

import re

PASSWORD_MIN_LENGTH = 8
PASSWORD_MAX_LENGTH = 128

_COMMON_PASSWORDS = frozenset(
    {
        "password",
        "password1",
        "password123",
        "12345678",
        "123456789",
        "qwerty123",
        "admin123",
        "letmein1",
        "welcome1",
        "changeme",
        "iloveyou",
    }
)

_COMPLEXITY_MESSAGE = (
    "Password must be at least 8 characters and include uppercase, lowercase, "
    "a number, and a symbol (e.g. ! @ # $)."
)


def validate_password_complexity(password: str) -> str:
    """Return password if valid; raise ValueError with a clear message if not."""
    if len(password) < PASSWORD_MIN_LENGTH:
        raise ValueError(_COMPLEXITY_MESSAGE)
    if len(password) > PASSWORD_MAX_LENGTH:
        raise ValueError(f"Password must be at most {PASSWORD_MAX_LENGTH} characters.")

    if not re.search(r"[A-Z]", password):
        raise ValueError(_COMPLEXITY_MESSAGE)
    if not re.search(r"[a-z]", password):
        raise ValueError(_COMPLEXITY_MESSAGE)
    if not re.search(r"\d", password):
        raise ValueError(_COMPLEXITY_MESSAGE)
    if not re.search(r"[^A-Za-z0-9]", password):
        raise ValueError(_COMPLEXITY_MESSAGE)

    normalized = password.strip().lower()
    if normalized in _COMMON_PASSWORDS:
        raise ValueError("Password is too common. Choose a stronger, unique password.")

    if len(set(password)) < 5:
        raise ValueError("Password is too simple. Use a mix of varied characters.")

    return password

File: frontend/src/utils/passwordPolicy.js
const MIN_LENGTH = 8
const MAX_LENGTH = 128

const COMMON_PASSWORDS = new Set([
  'password',
  'password1',
  'password123',
  '12345678',
  '123456789',
  'qwerty123',
  'admin123',
  'letmein1',
  'welcome1',
  'changeme',
  'iloveyou',
])

export const PASSWORD_REQUIREMENTS_HINT =
  'At least 8 characters with uppercase, lowercase, a number, and a symbol (e.g. ! @ # $).'

const COMPLEXITY_ERROR =
  'Password must be at least 8 characters and include uppercase, lowercase, a number, and a symbol (e.g. ! @ # $).'

export function validatePasswordStrength(password) {
  if (!password) {
    return { ok: false, error: 'Enter a new password.' }
  }
  if (password.length < MIN_LENGTH || password.length > MAX_LENGTH) {
    return { ok: false, error: COMPLEXITY_ERROR }
  }
  if (!/[A-Z]/.test(password)) {
    return { ok: false, error: COMPLEXITY_ERROR }
  }
  if (!/[a-z]/.test(password)) {
    return { ok: false, error: COMPLEXITY_ERROR }
  }
  if (!/\d/.test(password)) {
    return { ok: false, error: COMPLEXITY_ERROR }
  }
  if (!/[^A-Za-z0-9]/.test(password)) {
    return { ok: false, error: COMPLEXITY_ERROR }
  }

  const normalized = password.trim().toLowerCase()
  if (COMMON_PASSWORDS.has(normalized)) {
    return {
      ok: false,
      error: 'Password is too common. Choose a stronger, unique password.',
    }
  }

  if (new Set(password).size < 5) {
    return {
      ok: false,
      error: 'Password is too simple. Use a mix of varied characters.',
    }
  }

  return { ok: true }
}
